src/compiler/transformers/update-component-class.ts

```diff
@@ -3,19 +3,35 @@ import ts from 'typescript';
 import type * as d from '../../declarations';
 import { retrieveTsDecorators, retrieveTsModifiers } from './transform-utils';
 
+/**
+ * Transformation helper for updating how a Stencil component class is declared.
+ *
+ * Based on the output module type (CommonJS or ESM), the behavior is slightly different:
+ * - For CommonJS, the component class is left as is
+ * - For ESM, the component class is re-written as a variable statement
+ *
+ * @param transformOpts the options provided to TypeScript + Rollup for transforming the AST node
+ * @param classNode the node in the AST pertaining to the Stencil component class to transform
+ * @param heritageClauses a collection of heritage clauses associated with the provided class node
+ * @param members a collection of members attached to the provided class node
+ * @param moduleFile the Stencil intermediate representation associated with the provided class node
+ * @returns the updated component class declaration
+ */
```

Comment:
comprehension check with what's going on here:

We create the static getter in the `convertDecoratorsToStatic` transform that is passed to typescript and run through our initial pass through the typescript compiler.

We need to calculate and store that value (via the static getter) because we need that information later on in the compilation process, when we're marshalling components to pass them to Rollup in the output targets.

The reason we need to do the calculation in the `convertDecoratorsToStatic` is because if we defer until the separate transformers that we run as part of the output targets (where the call chain for this `updateComponentClass` function originates) then TypeScript has already made the transformation that we're trying to pre-empt, so we can no longer find a `static` property on our class and figure out if we need to do our own transformation.

Thus we need to hook into the compilation process early on:

https://github.com/ionic-team/stencil/blob/3091be91c009e3c98295221cb56980ab64fba61e/src/compiler/transpile/run-program.ts#L64-L67

then store then information so we can act on it later on when dealing with output target specific code.

Is that on the right track at least?

Reply:
Yep! Exactly - LMK if I didn't answer this/fill in additional details during our sync this morning

Reply:
Nope, all makes sense!